typing/datarepr.ml

```diff
@@ -107,10 +107,17 @@ let constructor_args ~current_unit priv cd_args cd_res path rep =
 let constructor_descrs ~current_unit ty_path decl cstrs rep =
   let ty_res = newgenconstr ty_path decl.type_params in
   let cstr_shapes_and_arg_jkinds =
-    match rep with
-    | Variant_extensible -> assert false
-    | Variant_boxed x -> x
-    | Variant_unboxed -> [| Constructor_uniform_value, [| decl.type_jkind |] |]
+    match rep, cstrs with
+    | Variant_extensible, _ -> assert false
+    | Variant_boxed x, _ -> x
+    | Variant_unboxed, [{ cd_args }] ->
+      begin match cd_args with
+      | Cstr_tuple [{ ca_jkind = jkind }]
+      | Cstr_record [{ ld_jkind = jkind }] ->
+        [| Constructor_uniform_value, [| jkind |] |]
+      | _ -> Misc.fatal_error "Multiple arguments in [@@unboxed] variant"
+      end
+    | _ -> Misc.fatal_error "Multiple constructors in [@@unboxed] variant"
```

Comment:
This change is the major change in this patch, actually, driving the motivation behind it all. I don't remember all the details, but `decl.type_jkind` comes, I think, from the original estimate of the kind. We'll update the final `type_jkind` to be correct, but this one won't be right, yet, and so we have to use the kind from the constructor arg.

That said, of course you're right about the catch-all patterns. Will fix.

Reply:
> I don't remember all the details, but decl.type_jkind comes, I think, from the original estimate of the kind. We'll update the final type_jkind to be correct, but this one won't be right, yet, and so we have to use the kind from the constructor arg.

I took a quick look through typedecl, and this doesn't yet makes sense to me. This function is reached by many paths, but I think the two relevant ones are probably the two calls to `add_types_to_env` in `Typedecl.transl_type_decl`. It's not obvious to me that the `ca_jkind`/`ld_jkind` should be more accurate than `decl.type_jkind` at either. At the first call site, we haven't yet reached `update_decl_jkind` yet so I think `ca_jkind`/`ld_jkind` will still be the "any" from `transl_labels` or `transl_types_gf`. At the second callsite, we're fully done checking the decl so I think the jkind of the field and the jkind of the overall type should be the same.

I'm sure I'm missing something basic here, but since it has eluded me it seems worth understanding and documenting with a comment on this function.

> This change is the major change in this patch, actually, driving the motivation behind it all.

If this change is the motivation, is there a test that demonstrates it?  Might be nice to point to from the comment.

Reply:
OK. I've added the CR.